File: stacks/parenthesis.java
//https://leetcode.com/problems/valid-parentheses/

import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack();
        int size = s.length();
        if (size == 1)
            return false;
        for (int i = 0; i < size; i++) {

            if (s.charAt(i) == '[' || s.charAt(i) == '(' || s.charAt(i) == '{') {
                st.push(s.charAt(i));
                continue;
            }

            if (!st.isEmpty() && st.peek() == '(') {
                if (s.charAt(i) == ')')
                    st.pop();
                continue;
            }
            if (!st.isEmpty() && st.peek() == '{') {
                if (s.charAt(i) == '}')
                    st.pop();
                continue;
            }
            if (!st.isEmpty() && st.peek() == '[') {
                if (s.charAt(i) == ']')
                    st.pop();
                continue;
            }
            if (st.isEmpty() && s.charAt(i) == ']' || s.charAt(i) == '}' || s.charAt(i) == ')') {
                return false;
            }

        }
        return st.size() == 0;
    }
}

File: stacks/parenthesis.py
#https://leetcode.com/problems/valid-parentheses/

class Solution:
    def isValid(self, s: str) -> bool:
        stack=[]
        for char in s:
            if char=='[' or char =='{' or char =='(':
                stack.append(char)
            else:
                if char==']':
                    if len(stack)==0 or not stack.remove(stack[len(stack)-1])=='[':
                        return False
                if char=='}':
                    if len(stack)==0 or not stack.remove(stack[len(stack)-1])=='{':
                        return False
                if char==')':
                    if len(stack)==0 or not stack.remove(stack[len(stack)-1])=='(':
                        return False
        return len(stack)==0
